metals/src/main/scala/scala/meta/internal/metals/OutlineFilesProvider.scala

```diff
@@ -0,0 +1,198 @@
+package scala.meta.internal.metals
+
+import java.util.Optional
+import java.util.concurrent.atomic.AtomicBoolean
+import java.{util => ju}
+
+import scala.collection.concurrent.TrieMap
+
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.io.AbsolutePath
+import scala.meta.pc.VirtualFileParams
+import scala.meta.pc.{OutlineFiles => JOutlineFiles}
+
+import ch.epfl.scala.bsp4j.BuildTargetIdentifier
+
+class OutlineFilesProvider(
+    buildTargets: BuildTargets,
+    buffers: Buffers,
+) {
+  private val outlineFiles =
+    new TrieMap[BuildTargetIdentifier, BuildTargetOutlineFilesProvider]()
+
+  def shouldRestartPc(
+      id: BuildTargetIdentifier,
+      reason: PcRestartReason,
+  ): Boolean = {
+    reason match {
+      case DidCompile(true) => true
+      case _ =>
+        outlineFiles.get(id) match {
+          case Some(provider) =>
+            // if it was never compiled successfully by the build server
+            // we don't restart pc not to lose information from outline compile
+            provider.wasSuccessfullyCompiledByBuildServer
+          case None => true
+        }
+    }
+  }
+
+  def onDidCompile(id: BuildTargetIdentifier, wasSuccessful: Boolean): Unit = {
+    outlineFiles.get(id) match {
+      case Some(provider) =>
+        if (wasSuccessful) provider.successfulCompilation()
+      case None =>
+        for {
+          scalaTarget <- buildTargets.scalaTarget(id)
+          // we don't perform outline compilation for Scala 3
+          if (!ScalaVersions.isScala3Version(scalaTarget.scalaVersion))
+        } outlineFiles.putIfAbsent(
+          id,
+          new BuildTargetOutlineFilesProvider(
+            buildTargets,
+            buffers,
+            id,
+            wasSuccessful,
+          ),
+        )
+    }
+  }
+
+  def didChange(id: String, path: AbsolutePath): Unit =
+    buildTargetId(id).foreach(didChange(_, path))
+
+  def didChange(id: BuildTargetIdentifier, path: AbsolutePath): Unit = {
+    for {
+      provider <- outlineFiles.get(id)
+    } provider.didChange(path)
+  }
+
+  def getOutlineFiles(id: String): Optional[JOutlineFiles] =
```

Comment:
It is `BuildTargetIdentifier`, I'm constructing it here.
```scala
  def getOutlineFiles(id: String): Optional[JOutlineFiles] =
    getOutlineFiles(buildTargetId(id))

  def getOutlineFiles(
      buildTargetId: Option[BuildTargetIdentifier]
  )
  ```
  It's just that it comes as a string from pc (can't be `BuildTargetIdentifier`, because pc interface doesn't depend on `bsp4j`) and it's less boilerplate to map it here instead of each place in `Compilers`.